README.md
[![Build Status](https://travis-ci.org/jean553/interface-tests-helpers.svg?branch=master)](https://travis-ci.org/jean553/interface-tests-helpers)

[![Current Crates.io Version](https://img.shields.io/crates/v/interface-tests-helpers.svg)](https://crates.io/crates/interface-tests-helpers)

# interface-tests-helpers

Routines for HTTP interface testing in Rust.

## Table of contents
- [Development](#development)
- [Generate documentation](#generate-documentation)
- [Usage](#usage)
- [Tests](#tests)

## Development

Build the development container.

```sh
vagrant up
```

Connect to the development container.

```sh
vagrant ssh
```

## Generate documentation

```sh
cargo rustdoc -- --document-private-items
```

## Usage

Cargo.toml:

```
[dev-dependencies]
interface-tests-helpers = "*"
```

```rust
extern crate interface_tests_helpers;

use interface_tests_helpers::{
    ClientHandler,
    ResponseHandler,
    HasBaseUrl,
};

use reqwest::{
    Client,
    Response,
};

use std::collections::HashMap;

impl HasBaseUrl for Client {

    /// Returns the service base URL.
    ///
    /// # Returns:
    ///
    /// the service base URL.
    fn get_base_url(&self) -> &str {
        "http://localhost:1234"
    }
}

trait ResourceHandler {

    fn post_resource(&self, json: &HashMap<&str, &str>) -> Response;
}

impl ResourceHandler for Client {

    /// Example of "per resource implementation" method.
    ///
    /// # Arguments:
    ///
    /// `json` - the json data to send
    fn post_resource(
        &self,
        json: &HashMap<&str, &str>,
    ) -> Response {

        return self.post_json(
            "/resource",
            json,
        );
    }
}

#[test]
fn test_post_resource() {

    let mut json: HashMap<&str, &str> = HashMap::new();
    json.insert("key", "value");

    let client = Client::new();
    let response = client.post_resource(&json);

    response.assert_201();
}
```

## Tests

One thread only must be used:

```rust
cargo test -- --test-threads=1
```


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.2.0] - 2019-06-13

### Changed
 - upgrade reqwest from 0.8.5 to 0.9.18 and mockito from 0.9.1 to 0.17.1

## [0.1.0] - 2018-03-24 

### Added
 - GET helper 
 - POST helper (with JSON content-type)
 - POST helper (with PlainText content-type)
 - PUT helper (with XML content-type)
 - PUT helper (with PlainText content-type)
 - Responses assertions for usual status codes


src/tests.rs
extern crate mockito;
extern crate reqwest;

#[cfg(test)]
mod tests {

    use ClientHandler;
    use ResponseHandler;

    use reqwest::{
        Client,
        Response,
    };

    use tests::mockito::mock;

    use std::collections::HashMap;

    trait HasBaseUrl {

        fn get_base_url(&self) -> &str;
    }

    impl HasBaseUrl for Client {

        /// Returns the service base URL.
        ///
        /// # Returns:
        ///
        /// the service base URL.
        fn get_base_url(&self) -> &str {
            "http://localhost:1234"
        }
    }

    trait ResourceHandler {

        fn post_json_resource(&self, json: &HashMap<&str, &str>) -> Response;

        fn post_body_resource(&self, body: &str) -> Response;

        fn get_resource(&self) -> Response;

        fn update_resource(&self, body: &str) -> Response;

        fn update_resource_text(&self, text: &str) -> Response;
    }

    impl ResourceHandler for Client {

        /// Example of "per resource implementation" method to post JSON.
        ///
        /// # Arguments:
        ///
        /// `json` - the json data to send
        ///
        /// # Returns:
        ///
        /// reqwest response
        fn post_json_resource(
            &self,
            json: &HashMap<&str, &str>,
        ) -> Response {

            self.post_json(
                &format!("{}/resource", self.get_base_url()),
                json,
            )
        }

        /// Example of "per resource implementation" method to post raw body.
        ///
        /// # Arguments:
        ///
        /// `body` - the raw body to send
        ///
        /// # Returns:
        ///
        /// reqwest response
        fn post_body_resource(
            &self,
            body: &str,
        ) -> Response {

            self.post_body(
                &format!("{}/resource", self.get_base_url()),
                body,
            )
        }

        /// Example of "per resource implementation" method to get URL.
        ///
        /// # Returns:
        ///
        /// reqwest response
        fn get_resource(&self) -> Response {

            self.get_url(&format!("{}/resource", self.get_base_url()))
        }

        /// Example of "per resource implementation" method to PUT content.
        ///
        /// # Returns:
        ///
        /// reqwest response
        fn update_resource(
            &self,
            body: &str,
        ) -> Response {

            self.put_xml(
                &format!("{}/resource", self.get_base_url()),
                body,
            )
        }

        /// Example of "per resource implementation" method to PUT text content.
        ///
        /// # Returns:
        ///
        /// reqwest response
        fn update_resource_text(
            &self,
            body: &str,
        ) -> Response {

            self.put_text(
                &format!("{}/resource", self.get_base_url()),
                body,
            )
        }
    }

    #[test]
    fn test_post_json_returns_201() {

        const API: &str = "/resource";
        let _m = mock("POST", API)
            .with_status(201)
            .with_body("OK")
            .create();

        let mut json: HashMap<&str, &str> = HashMap::new();
        json.insert("key", "value");

        let client = Client::new();
        let response = client.post_json_resource(&json);

        response.assert_201();
    }

    #[test]
    fn test_post_json_returns_400() {

        const API: &str = "/resource";
        let _m = mock("POST", API)
            .with_status(400)
            .create();

        let json: HashMap<&str, &str> = HashMap::new();

        let client = Client::new();
        let response = client.post_json_resource(&json);

        response.assert_400();
    }

    #[test]
    fn test_post_json_returns_409() {

        const API: &str = "/resource";
        let _m = mock("POST", API)
            .with_status(409)
            .create();

        let mut json: HashMap<&str, &str> = HashMap::new();
        json.insert("key", "value");

        let client = Client::new();
        let response = client.post_json_resource(&json);

        response.assert_409();
    }

    #[test]
    fn test_post_body_returns_201() {

        const API: &str = "/resource";
        let _m = mock("POST", API)
            .with_status(201)
            .with_body("OK")
            .create();

        let client = Client::new();
        let response = client.post_body_resource("raw body");

        response.assert_201();
    }

    #[test]
    fn test_get_returns_200() {

        const API: &str = "/resource";
        let _m = mock("GET", API)
            .with_status(200)
            .create();

        let client = Client::new();
        let response = client.get_resource();

        response.assert_200();
    }

    #[test]
    fn test_assert_204() {

        const API: &str = "/resource";
        let _m = mock("POST", API)
            .with_status(204)
            .create();

        let client = Client::new();
        let response = client.post_body_resource("raw body");

        response.assert_204();
    }

    #[test]
    fn test_assert_401() {

        const API: &str = "/resource";
        let _m = mock("GET", API)
            .with_status(401)
            .create();

        let client = Client::new();
        let response = client.get_resource();

        response.assert_401();
    }

    #[test]
    fn test_assert_403() {

        const API: &str = "/resource";
        let _m = mock("GET", API)
            .with_status(403)
            .create();

        let client = Client::new();
        let response = client.get_resource();

        response.assert_403();
    }

    #[test]
    fn test_assert_404() {

        const API: &str = "/resource";
        let _m = mock("GET", API)
            .with_status(404)
            .create();

        let client = Client::new();
        let response = client.get_resource();

        response.assert_404();
    }

    #[test]
    fn test_assert_500() {

        const API: &str = "/resource";
        let _m = mock("GET", API)
            .with_status(500)
            .create();

        let client = Client::new();
        let response = client.get_resource();

        response.assert_500();
    }

    #[test]
    fn test_put_xml() {

        const API: &str = "/resource";
        let _m = mock("PUT", API)
            .with_status(200)
            .create();

        let client = Client::new();

        let xml = "<key>value</key>";
        let response = client.update_resource(&xml);

        response.assert_200();
    }

    #[test]
    fn test_put_text() {

        const API: &str = "/resource";
        let _m = mock("PUT", API)
            .with_status(200)
            .create();

        let client = Client::new();
        let response = client.update_resource_text("text");

        response.assert_200();
    }
}


src/lib.rs
//! Tests helpers to test HTTP interfaces.

extern crate reqwest;

use reqwest::{
    Client,
    Response,
    StatusCode,
};

use std::collections::HashMap;

pub trait HasBaseUrl {

    fn get_base_url(&self) -> &str;
}

pub trait ClientHandler {

    fn get_url(&self, url: &str) -> Response;

    fn post_json(&self, url: &str, json: &HashMap<&str, &str>) -> Response;

    fn post_body(&self, url: &str, body: &str) -> Response;

    fn put_xml(&self, url: &str, body: &str) -> Response;

    fn put_text(&self, url: &str, text: &str) -> Response;
}

pub trait ResponseHandler {

    fn assert_200(&self);

    fn assert_201(&self);

    fn assert_204(&self);

    fn assert_400(&self);

    fn assert_401(&self);

    fn assert_403(&self);

    fn assert_404(&self);

    fn assert_409(&self);

    fn assert_500(&self);
}

impl ClientHandler for Client {

    /// Perform a GET request without parameter
    ///
    /// # Args:
    ///
    /// `url` - the suffix of the URL
    fn get_url(
        &self,
        url: &str,
    ) -> Response {

        self.get(url)
            .send()
            .unwrap()
    }

    /// Perform a POST request to send JSON and stores its result
    ///
    /// # Args:
    ///
    /// `url` - the suffix of the URL
    /// `json` - the json data to send
    fn post_json(
        &self,
        url: &str,
        json: &HashMap<&str, &str>,
    ) -> Response {

        self.post(url)
            .json(json)
            .send()
            .unwrap()
    }

    /// Perform a POST request to send a raw body and stores its result
    ///
    /// # Args:
    ///
    /// `url` - the suffix of the URL
    /// `body` - the body data to send (String format)
    fn post_body(
        &self,
        url: &str,
        body: &str,
    ) -> Response {

        self.post(url)
            .body(body.to_string())
            .header(
                reqwest::header::CONTENT_TYPE,
                "text/plain"
            )
            .send()
            .unwrap()
    }

    /// Perform a PUT request to send a XML body and returns its result
    ///
    /// # Args:
    ///
    /// `url` - the suffix of the URL
    /// `body` - the body data to send (XML format)
    fn put_xml(
        &self,
        url: &str,
        body: &str,
    ) -> Response {

        self.put(url)
            .body(body.to_string())
            .header(
                reqwest::header::CONTENT_TYPE,
                "application/xml"
            )
            .send()
            .unwrap()
    }

    /// Perform a PUT request to send a text body and returns its result
    ///
    /// # Args:
    ///
    /// `url` - the suffix of the URL
    /// `text` - the text to upload
    fn put_text(
        &self,
        url: &str,
        text: &str,
    ) -> Response {

        self.put(url)
            .body(text.to_string())
            .header(
                reqwest::header::CONTENT_TYPE,
                "text/plain"
            )
            .send()
            .unwrap()
    }
}

impl ResponseHandler for Response {

    /// Assertion that checks the response status code is 200
    fn assert_200(&self) {

        assert_eq!(
            self.status(),
            StatusCode::OK,
        );
    }

    /// Assertion that checks the response status code is 201
    fn assert_201(&self) {

        assert_eq!(
            self.status(),
            StatusCode::CREATED,
        );
    }

    /// Assertion that checks the response status code is 204
    fn assert_204(&self) {

        assert_eq!(
            self.status(),
            StatusCode::NO_CONTENT,
        );
    }

    /// Assertion that checks the response status code is 400
    fn assert_400(&self) {

        assert_eq!(
            self.status(),
            StatusCode::BAD_REQUEST,
        );
    }

    /// Assertion that checks the response status code is 401
    fn assert_401(&self) {

        assert_eq!(
            self.status(),
            StatusCode::UNAUTHORIZED,
        );
    }

    /// Assertion that checks the response status code is 403
    fn assert_403(&self) {

        assert_eq!(
            self.status(),
            StatusCode::FORBIDDEN,
        );
    }

    /// Assertion that checks the response status code is 404
    fn assert_404(&self) {

        assert_eq!(
            self.status(),
            StatusCode::NOT_FOUND,
        );
    }

    /// Assertion that checks the response status code is 409
    fn assert_409(&self) {

        assert_eq!(
            self.status(),
            StatusCode::CONFLICT,
        );
    }

    /// Assertion that checks the response status code is 500
    fn assert_500(&self) {

        assert_eq!(
            self.status(),
            StatusCode::INTERNAL_SERVER_ERROR,
        );
    }
}

#[cfg(test)]
mod tests;


src/tests_post.rs
extern crate mockito;
extern crate reqwest;

#[cfg(test)]
mod tests {

    use ClientHandler;
    use ResponseHandler;

    use reqwest::{
        Client,
        Response,
    };

    use tests_post::mockito::mock;

    use std::collections::HashMap;

    trait HasBaseUrl {

        fn get_base_url(&self) -> &str;
    }

    impl HasBaseUrl for Client {

        /// Returns the service base URL.
        ///
        /// # Returns:
        ///
        /// the service base URL.
        fn get_base_url(&self) -> &str {
            "http://localhost:1234"
        }
    }

    trait ResourceHandler {

        fn post_json_resource(&self, json: &HashMap<&str, &str>) -> Response;

        fn post_body_resource(&self, body: &str) -> Response;
    }

    impl ResourceHandler for Client {

        /// Example of "per resource implementation" method to post JSON.
        ///
        /// # Arguments:
        ///
        /// `json` - the json data to send
        fn post_json_resource(
            &self,
            json: &HashMap<&str, &str>,
        ) -> Response {

            self.post_json(
                &format!("{}/resource", self.get_base_url()),
                json,
            )
        }

        /// Example of "per resource implementation" method to post raw body.
        ///
        /// # Arguments:
        ///
        /// `body` - the raw body to send
        fn post_body_resource(
            &self,
            body: &str,
        ) -> Response {

            self.post_body(
                &format!("{}/resource", self.get_base_url()),
                body,
            )
        }
    }

    #[test]
    fn test_post_json_returns_201() {

        const API: &str = "/resource";
        let _m = mock("POST", API)
            .with_status(201)
            .with_body("OK")
            .create();

        let mut json: HashMap<&str, &str> = HashMap::new();
        json.insert("key", "value");

        let client = Client::new();
        let response = client.post_json_resource(&json);

        response.assert_201();
    }

    #[test]
    fn test_post_json_returns_400() {

        const API: &str = "/resource";
        let _m = mock("POST", API)
            .with_status(400)
            .create();

        let json: HashMap<&str, &str> = HashMap::new();

        let client = Client::new();
        let response = client.post_json_resource(&json);

        response.assert_400();
    }

    #[test]
    fn test_post_json_returns_409() {

        const API: &str = "/resource";
        let _m = mock("POST", API)
            .with_status(409)
            .create();

        let mut json: HashMap<&str, &str> = HashMap::new();
        json.insert("key", "value");

        let client = Client::new();
        let response = client.post_json_resource(&json);

        response.assert_409();
    }
}
